msrestazure/azure_local_creds_prober.py

```diff
@@ -0,0 +1,280 @@
+                                     AZURE_PUBLIC_CLOUD, AZURE_US_GOV_CLOUD)
+
+_LOGGER = logging.getLogger(__name__)
+
+#pylint: disable=too-few-public-methods,missing-docstring
+
+class CredsProber:
+    '''
+    Prober base class
+    '''
+    def __init__(self, resource):
+        self.enabled = True
+        self.resource = resource
+
+    def find_cloud_env(self, cloud_name):
+        envs = [AZURE_PUBLIC_CLOUD, AZURE_CHINA_CLOUD, AZURE_GERMAN_CLOUD, AZURE_US_GOV_CLOUD]
+        if cloud_name:
+            env = next((e for e in envs if cloud_name.lower() == e.name.lower()), None)
+            if not env:
+                raise ValueError('"{}" is not a supported cloud environment by Python'
+                                 ' SDK'.format(env))
+        else:
+            env = AZURE_PUBLIC_CLOUD
+        return env
+
+    def probe_subscription(self, creds):  # pylint: disable=no-self-use
+        subscription_id = None
+        try:
+            from azure.mgmt.resource.subscriptions import SubscriptionClient
+            subscriptions = list(SubscriptionClient(creds).subscriptions.list())
+            if subscriptions:
+                subscription_ids = [s.id.split('/')[-1] for s in subscriptions]
+                subscription_id = subscription_ids[0]
+                _LOGGER.warning('Found subscription "%s" to use', subscription_ids[0])
+                if len(subscription_ids) > 1:
+                    _LOGGER.warning('You also have accesses to a few other subscriptions "%S".'
+                                    ' You can supply subscription_id on invoking probers')
+        except ImportError:  # should be rare
+            _LOGGER.warning('Failed to load azure.mgmt.resource.subscriptions to find the default'
+                            ' subscription. If this is expected, supply subscription_id on creating'
+                            ' the probe object')
+        return subscription_id
+
+
+class ManagedServiceIdentityProber(CredsProber):
+    '''
+    Detect managed service identity coming with VM, VM scaleset, and App Service
+    '''
+    def probe(self):
+        if not self.enabled:
+            return None
+        try:
+            creds = MSIAuthentication()
+             # MSI is not yet supported in sovereigns
+            setattr(creds, 'cloud_environment', AZURE_PUBLIC_CLOUD)
+            creds.resource = self.resource or AZURE_PUBLIC_CLOUD.endpoints.management
+            _LOGGER.warning('Managed system identity was detected')
+            return creds
+        except (requests.exceptions.ConnectionError, requests.exceptions.HTTPError):
+            return None
+
+
+class ConnectionStrEnvProber(CredsProber):
+    '''
+    Detect environment variable AZURE_CONNECTION_STRING
+    '''
+    def probe(self):
+        creds = None
+        if self.enabled and os.environ.get('AZURE_CONNECTION_STRING'):
+            auth_info = json.loads(os.environ.get('AZURE_CONNECTION_STRING'))
```

Comment:
Is the connection string JSON valid? I thought it was something like:
AZURE_CONNECTION_STRING="client_id=1234;client_secret=2345;tenant_id=3456;subscription_id=4567;"